### PR TITLE
Added constructor for multidimensional tensors with params long[]

### LINKED ISSUE #9: Add a `Tensor (long [] dims)` constructor as a convenience ctor that creates 1D..4D tensors.

### PR DESCRIPTION
**Pull request by interesaaat**
Fix #9

### CODE REVIEW

File: TorchSharp/TypeGeneration.tt [hunk 1]
```diff
@@ -163,96 +163,76 @@ foreach (var type in new [] {
...
-        public <#=tname#>Tensor (long size0, long size1, long size2, long size3)
-        {
-            handle = TH<#=tname#>Tensor_newWithSize4d (size0, size1, size2, size3);
+        ///    Creates a 1-4D tensor of the specified size(s).
+        /// </summary>    
+        /// <param name="dims">Sizes for the dimensions.</param>
+
+        public <#=tname#>Tensor (params long[] dims)
+        {
+			switch (dims.Length)
```
Comment (markusweimer) [inline TorchSharp/TypeGeneration.tt]: Doesn't this `switch` slow things down compared to the explicit overloads? Why not keep both?

Comment (interesaaat) [reply TorchSharp/TypeGeneration.tt]: Probably, but is 1 call (not data dependent). I can add back the previous constructors, WDYT?

Comment (migueldeicaza) [reply TorchSharp/TypeGeneration.tt]: We should keep both.

**Comment by migueldeicaza** [comment]
I will hold back some changes I have to cleanup the type hierarchy until this gets in to not make your life harder.

File: TorchSharp/TypeGeneration.tt (hunk 1 above)
Comment (interesaaat) [reply TorchSharp/TypeGeneration.tt]: Fixed.

**Comment by interesaaat** [comment]
> I will hold back some changes I have to cleanup the type hierarchy until this gets in to not make your life harder.

Can we have a generic `TorchSharp<T>` abstract class all the `<#=tname#>Tensor`s can inherit from? This will make my life easier on the SNT bindings side.

File: TorchSharp/TypeGeneration.tt [hunk 2]
```diff
@@ -255,6 +257,34 @@ foreach (var type in new [] {
...
+		            handle = TH<#=tname#>Tensor_newWithSize2d (dims[0], dims[1]);
+		            break;
+	            case 3: 
+		            handle = TH<#=tname#>Tensor_newWithSize3d (dims[0], dims[1], dims[2]);
+		            break;
+	            case 4: 
+		            handle = TH<#=tname#>Tensor_newWithSize4d (dims[0], dims[1], dims[2], dims[3]);
+		            break;
+	            default:
+		            throw new ArgumentOutOfRangeException(nameof(dims), "Maximum number of dimensions for tensor creation is 4.");
```
[the remark this answers is not in the document]
Comment (interesaaat) [reply TorchSharp/TypeGeneration.tt]: That's a good point. Is there a `reshape` function in TH? I am not able to find it.

**Comment by markusweimer** [comment]
@interesaaat Can you bring this PR into mergable status again? Also, is it ready for review?

**Comment by interesaaat** [comment]
> @interesaaat Can you bring this PR into mergable status again? Also, is it ready for review?

Updated and  good for review.

File: TorchSharp/TypeGeneration.cs [hunk 3]
```diff
@@ -3581,7 +3637,7 @@ public long GetTensorDimension (int dim)
             get {
                     var dims = new long [Dimensions];
                     for (int i = 0; i < dims.Length; i++)
-                            dims [i] = (long)GetTensorDimension (i);
```
Comment (markusweimer) [inline TorchSharp/TypeGeneration.cs]: What changed that allowed us to drop the cast?

File: TorchSharp/TypeGeneration.tt (hunk 2 above)
[the remark this answers is not in the document]
Comment (motus) [reply TorchSharp/TypeGeneration.tt]: @interesaaat oh never mind then. we can commit it as it is, and then add `reshape` instead of an exception in some subsequent PR wehn we have such functionality

[the remark this answers is not in the document]
Comment (interesaaat) [reply TorchSharp/TypeGeneration.tt]: Ok, marking as resolved. We will have to remember to fill an issue if we want to do the `reshape` thing.

File: TorchSharp/TypeGeneration.cs (hunk 3 above)
Comment (interesaaat) [reply TorchSharp/TypeGeneration.cs]: Nothing, in fact the cast is superfluous.

File: TorchSharp/TypeGeneration.cs [hunk 4]
```diff
@@ -238,6 +238,34 @@ public ByteTensor (long size0, long size1, long size2, long size3)
...
+                    break;
+                case 2: 
+                    handle = THByteTensor_newWithSize2d (dims[0], dims[1]);
+		            break;
+                case 3: 
+                    handle = THByteTensor_newWithSize3d (dims[0], dims[1], dims[2]);
+		            break;
+                case 4: 
+                    handle = THByteTensor_newWithSize4d (dims[0], dims[1], dims[2], dims[3]);
+		            break;
```
[the remark this answers is not in the document]
Comment (interesaaat) [reply TorchSharp/TypeGeneration.cs]: Mmm I thought I solved this. Let me try again.

[the remark this answers is not in the document]
Comment (interesaaat) [reply TorchSharp/TypeGeneration.cs]: Fixed

**Comment by motus** [comment]
Merged in f53a2e97e23f8314ef5ceee56355621340b3bee3

**Comment by motus** [comment]
oops, looks like we did the same thing simultaneously :) I've tested and merged this PR manually
